File: flutter/sample_ui/test/state/use_smileid_sample_top_app_bar_semantics_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:sample_ui/sample_ui.dart';

/// The app bar's controls have to be reachable one at a time, which a golden cannot show.
///
/// Found when the detail page became the app bar's first caller: with no trailing action the whole
/// row collapsed into one node labelled 'Back\nTitle', flagged as a button. A screen reader
/// announced the title as part of the back control, and neither could be reached alone.
void main() {
  /// Every labelled node in the tree, with the flags that decide how it is announced.
  List<({String label, bool button, bool header})> labelledNodes(
    WidgetTester tester,
  ) {
    final List<({String label, bool button, bool header})> found =
        <({String label, bool button, bool header})>[];
    void walk(SemanticsNode node) {
      if (node.label.isNotEmpty) {
        found.add((
          label: node.label,
          button: node.flagsCollection.isButton,
          header: node.flagsCollection.isHeader,
        ));
      }
      node.visitChildren((SemanticsNode child) {
        walk(child);
        return true;
      });
    }

    // The non-deprecated `rootPipelineOwner` holds no semantics owner in a widget test, so this
    // stays on the deprecated one until a replacement exists that actually works here.
    // ignore: deprecated_member_use
    walk(tester.binding.pipelineOwner.semanticsOwner!.rootSemanticsNode!);
    return found;
  }

  Future<List<({String label, bool button, bool header})>> pumpBar(
    WidgetTester tester, {
    Widget? action,
  }) async {
    // Disposed INLINE, not in a tear-down: flutter_test verifies no handle is live before its
    // tear-downs run, so a deferred dispose fails the test it was meant to clean up after. The
    // nodes are plain records by the time this returns, so nothing needs semantics to stay on.
    final SemanticsHandle semantics = tester.ensureSemantics();
    await tester.pumpWidget(
      MaterialApp(
        theme: UseSmileIDSampleTheme.light(),
        home: UseSmileIDSampleTopAppBar(
          title: 'Verification details',
          onBack: () {},
          action: action,
        ),
      ),
    );
    await tester.pumpAndSettle();
    final List<({String label, bool button, bool header})> nodes =
        labelledNodes(tester);
    semantics.dispose();
    return nodes;
  }

  // The configuration that was broken, and the one most screens use: back and a title, no action.
  testWidgets('with no action the back control and the title stay apart', (
    WidgetTester tester,
  ) async {
    final List<({String label, bool button, bool header})> nodes =
        await pumpBar(tester);

    expect(nodes.map((n) => n.label), <String>['Back', 'Verification details']);
    expect(nodes.first.button, isTrue);
    expect(nodes.last.header, isTrue, reason: 'the header is the title');
    expect(
      nodes.last.button,
      isFalse,
      reason: 'a title announced as a button offers an action it does not have',
    );
  });

  testWidgets('with an action all three are separately reachable', (
    WidgetTester tester,
  ) async {
    final List<({String label, bool button, bool header})> nodes =
        await pumpBar(
          tester,
          action: UseSmileIDSampleTopAppBarButton(
            semanticLabel: 'Hide verification from the app list',
            onTap: () {},
            glyph: UseSmileIDSampleGlyphs.trash,
          ),
        );

    expect(nodes.map((n) => n.label), <String>[
      'Back',
      'Verification details',
      'Hide verification from the app list',
    ]);
  });

  // The symptom the merge produced, asserted directly: no node may carry two controls' words.
  testWidgets('no node carries more than one label', (
    WidgetTester tester,
  ) async {
    for (final List<({String label, bool button, bool header})> nodes
        in <List<({String label, bool button, bool header})>>[
          await pumpBar(tester),
          await pumpBar(
            tester,
            action: UseSmileIDSampleTopAppBarButton(
              semanticLabel: 'Hide',
              onTap: () {},
              glyph: UseSmileIDSampleGlyphs.trash,
            ),
          ),
        ]) {
      for (final ({String label, bool button, bool header}) node in nodes) {
        expect(
          node.label,
          isNot(contains('\n')),
          reason: 'merged: ${node.label.replaceAll('\n', ' + ')}',
        );
      }
    }
  });
}

File: android/sample-ui/src/test/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleTopAppBarSemanticsTest.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.runtime.Composable
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.SemanticsNode
import androidx.compose.ui.semantics.SemanticsProperties
import androidx.compose.ui.semantics.getOrNull
import androidx.compose.ui.test.ExperimentalTestApi
import androidx.compose.ui.test.onRoot
import androidx.compose.ui.test.v2.runComposeUiTest
import com.usesmileid.sampleapps.ui.golden.ROBOLECTRIC_SDK
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/** The app bar's controls have to be reachable one at a time, which a screenshot cannot show. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [ROBOLECTRIC_SDK], qualifiers = "w411dp-h891dp-xhdpi")
@OptIn(ExperimentalTestApi::class)
class UseSmileIDSampleTopAppBarSemanticsTest {

    private data class Announced(val label: String, val button: Boolean, val header: Boolean)

    private fun bar(action: (@Composable () -> Unit)?): List<Announced> {
        lateinit var announced: List<Announced>
        runComposeUiTest {
            setContent {
                UseSmileIDSampleTheme(darkTheme = false) {
                    UseSmileIDSampleTopAppBar(title = TITLE, onBack = {}, action = action)
                }
            }
            announced = onRoot().fetchSemanticsNode().labelled()
        }
        return announced
    }

    private fun SemanticsNode.labelled(): List<Announced> {
        val found = mutableListOf<Announced>()
        fun walk(node: SemanticsNode) {
            val label = node.config.getOrNull(SemanticsProperties.ContentDescription)?.firstOrNull()
                ?: node.config.getOrNull(SemanticsProperties.Text)?.firstOrNull()?.text
            if (!label.isNullOrEmpty()) {
                found += Announced(
                    label = label,
                    button = node.config.getOrNull(SemanticsProperties.Role) == Role.Button,
                    header = node.config.getOrNull(SemanticsProperties.Heading) != null,
                )
            }
            node.children.forEach(::walk)
        }
        walk(this)
        return found
    }

    // The configuration most screens use, and the one a merge collapses first.
    @Test
    fun `with no action the back control and the title stay apart`() {
        val nodes = bar(action = null)

        assertEquals(listOf(BACK, TITLE), nodes.map { it.label })
        assertEquals("the back control is a button", true, nodes.first().button)
        assertEquals("the header is the title", true, nodes.last().header)
        assertEquals(
            "a title announced as a button offers an action it does not have",
            false,
            nodes.last().button,
        )
    }

    @Test
    fun `with an action all three are separately reachable`() {
        val nodes = bar(action = { UseSmileIDSampleTopAppBarButton(DELETE, onClick = {}) { TrashGlyph(it) } })

        assertEquals(listOf(BACK, TITLE, DELETE), nodes.map { it.label })
    }

    // The symptom a merge produces, asserted without naming the strings it would join.
    @Test
    fun `no node carries more than one label`() {
        val merged = listOf(
            bar(action = null),
            bar(action = { UseSmileIDSampleTopAppBarButton(DELETE, onClick = {}) { TrashGlyph(it) } }),
        ).flatten().filter { it.label.contains('\n') }.map { it.label.replace("\n", " + ") }

        assertEquals(emptyList<String>(), merged)
    }

    private companion object {
        const val BACK = "Back"
        const val TITLE = "Verification details"
        const val DELETE = "Hide verification from the app list"
    }
}
